### ISSUE TITLE
Zwift Bridge on Windows

### ISSUE THREAD

**Issue by cagnulein** [cagnulein/qdomyos-zwift#39]
Connection to the treadmill or bike should work, but the bridge it's not possible ATM because the peripheral feature is not yet implemented on qt 5.15.1

https://doc.qt.io/qt-5/qtbluetooth-index.html

**Comment by cagnulein** [comment]
i've succesfully built it using Qt VS Tools on Visual Studio 2017 with 
![Immagine](https://user-images.githubusercontent.com/48216/97455933-a90d0d80-1938-11eb-932c-053929cfefe4.png)
Qt MSVC 2015

**Comment by cagnulein** [comment]
https://forum.qt.io/topic/120364/bluetooth-le-peripheral-on-windows-uwp
asked support on Qt community

**Comment by cagnulein** [comment]
an idea to run this software *NOW* on windows is to use Virtualbox or similar with linux passing through the bluetooth adapter. I guess it should work flawlessy

**Comment by cagnulein** [comment]
via VirtualBox using the passthrough function for the bluetooth dongle works!
I have an image to try if anyone is interested (3 GB)

**Comment by cagnulein** [comment]
Ok these are the steps in order to have qdomyos-zwift running on Windows

1. Download Virtualbox from here https://download.virtualbox.org/virtualbox/6.1.16/VirtualBox-6.1.16-140961-Win.exe
2. Install Virtualbox of course
3. Download the qdomyos-image (it's 3GB, so have patience) http://80.211.67.253:8080/qdomyos-zwift.ova
4. Double click on the ova file downloaded
5. Attach to the running VM, your bluetooth adapter. In the image below, this is my setup on my dell notebook

![Immagine](https://user-images.githubusercontent.com/48216/98226049-cc961080-1f55-11eb-9a55-9b579f5fa8ff.png)

**Comment by rickoneeleven** [comment]
add to instructions you need extension pack from https://www.virtualbox.org/wiki/Downloads as per 
![VirtualBox_jh9vdPVDXk](https://user-images.githubusercontent.com/8156878/98383850-ecf0c880-2044-11eb-8f25-70732d68d98d.png)

**Comment by cagnulein** [comment]
done

**Comment by r3sonant** [comment]
Hello. Roberto sent me here for the VM setup guide.

No probs with getting the image up and running, but should qdomyos-zwift just see the bike when the bluetooth is connected? My bluetooth adapter is definitely connecting to the VM as my BT mouse stops working, I'm not seeing anything in qdomyos-zwift though.

Thanks.

**Comment by cagnulein** [comment]
Hi @r3sonant ok let's do in this way.
Close the qdomyos interface that opens at the startup of the VM.
Open a shell and type

<pre>sudo bash
cd Desktop/qdomyos-zwift/src/
./qdomyos-zwift -qml</pre>

and share the logs that this command produces here.
Thanks

**Comment by r3sonant** [comment]
[Intel BT.txt](https://github.com/cagnulein/qdomyos-zwift/files/7518935/Intel.BT.txt)
[BTDongle.txt](https://github.com/cagnulein/qdomyos-zwift/files/7518936/BTDongle.txt)
Hi @cagnulein 
Here's the logs.
I tried with the onboard Bluetooth and a Bluetooth 4 dongle I have. Both were showing as UP RUNNING in Linux when running hciconfig -a and the Bluetooth icon was flashing in the app and it was looking for the bike but it wasn't round.

**Comment by cagnulein** [comment]
@r3sonant oh got it! in the image there was selected a manual device setting (advanced setting) set to "iconsole bike" :D just put to disabled and restart QZ. Let me know

**Comment by r3sonant** [comment]
[Intel BT 2.txt](https://github.com/cagnulein/qdomyos-zwift/files/7522791/Intel.BT.2.txt)
e.
All very strange. I fired up the VM, disabled the bike selection in the advanced settings as you suggested. Restarted the app. The bike connected (It shows Bluetooth on the console when it's connected) but that seemed to crash the app.

I restarted and tried again, but it did the same, so I restarted the VM. Now the app doesn't crash but the bike doesn't connect.

The logs I have attached are from the last attempt when the bike didn't connect.

I also tried connecting from my phone, but that didn't connect either. So I tried on my tablet. That connected fine, so I fired Zwift up on my laptop. It saw my tablet and seemed to work.

So I cleared the app storage on my phone. It then connected to the bike. Zwift on the tablet saw the phone and connected, then said not connected.

I seem to be going round in circles. Something will partly work, but the next step won't work. Then the first step won't work.

All very confusing.

Thanks for your help so far.

**Comment by cagnulein** [comment]
@r3sonant the error in the log is 

`Thu Nov 11 14:17:53 2021 1636658273184 Debug:   Bluez 5 detected.
qt.bluetooth.bluez: Bluez 5 detected.
Thu Nov 11 14:17:53 2021 1636658273189 Debug: bluetooth.cpp void bluetooth::debug(const QString&) "no bluetooth dongle found!"`

I guess it's something about your dongle in the VM

about the app issue on the phone, i need the debug log also (you should open another ticket for that one ;) )

**Comment by r3sonant** [comment]
@cagnulein 

I tried with the onboard Bluetooth and a dongle but got the same result.

Where should I open another ticket for the phone?

Thanks.

**Comment by cagnulein** [comment]
@r3sonant but in the log that you did 2 days ago the dongles are well recognized. if the error is always the dongle issue. try to run "bluetoothctl" command you should be able to double check your blueooth configuration in linux.

> Where should I open another ticket for the phone?

here in this github project :) attaching the debug log from your phone with the issue

**Comment by r3sonant** [comment]
I have created a new ticket and attached the log file.

Thanks.

**Comment by jonm488** [comment]
@cagnulein I'm following up from an e-mail inquiry. Is this method still valid?

My set-up will be Dell 20" AIO w/ onboard bluetooth. Extra bluetooth dongle for VM + QZ. In the email you mentioned 2 bt dongles, but are onboard and dongle bt adequate?

**Comment by cagnulein** [comment]
hi @jonm488 yes this method is still valid.
Yes it should be adequate.
If you have to buy a dongle, buy the one that i'm advising in the wiki.
Let me know how it will go

**Comment by jonm488** [comment]
Thank you. I already purchased the newer version (TP-Link UB500) of your suggestion before finding it in the wiki. I figured for $3 more it was worth have BT 5.0. Hopefully it works just as well!

For reference to anybody who comes across this thread, the suggested dongle (TP-Link UB400) is at the bottom of [this wiki link](https://github.com/cagnulein/qdomyos-zwift/wiki/I-can't-connect-QZ-to-Zwift). I'll follow-up on how I do with the UB500.

**Comment by cagnulein** [comment]
@jonm488 the UB500 could be worst, that's why i'm still suggesting the previous model as @Cicciofritz79 could confirm

**Comment by jonm488** [comment]
Thanks @cagnulein. I just ordered the UB400 as well.

**Comment by cagnulein** [comment]
also it's possible to run android apps to windows 11 now https://tech.slashdot.org/story/22/04/18/1846246/android-apps-on-windows-11-review?utm_source=rss1.0mainlinkanon&utm_medium=feed

**Comment by cagnulein** [comment]
first windows binary with DIRCON #756 
you need:
- install qt 5.15.2
- run this .exe with "-qml" argument

I know that don't work:
- strava
- email
They are both easy to fix, i just want to test this out

**Comment by jeremy10tie** [comment]
> also it's possible to run android apps to windows 11 now https://tech.slashdot.org/story/22/04/18/1846246/android-apps-on-windows-11-review?utm_source=rss1.0mainlinkanon&utm_medium=feed

Hello, I had tested Windows 11 at some point the functionality of android applications (app store), but qz , managed to open but found nothing because Bluetooth was not working, since I am back to Windows 10, the time Windows 11 is still improving.

**Comment by cagnulein** [comment]
hi @jeremy10tie, check my previous comment about windows version ;)

**Comment by cagnulein** [comment]
update with all the qt libraries! https://github.com/cagnulein/qdomyos-zwift/suites/6203708054/artifacts/218893518
start it -qml arguments!

**Comment by soccer1886** [comment]
I'm getting libgcc_s_seh-1.dll & libstdc++-6.dll not found running the windows bin.

**Comment by cagnulein** [comment]
@soccer1886 thanks, I will try to check if I can fix it right now

**Comment by cagnulein** [comment]
@soccer1886 it's building, let's see when it finish https://github.com/cagnulein/qdomyos-zwift/runs/6163377940?check_suite_focus=true

**Comment by cagnulein** [comment]
@soccer1886 did you try this https://github.com/cagnulein/qdomyos-zwift/suites/6257070145/artifacts/222896918 ?

**Comment by soccer1886** [comment]
That works now but I have no idea how to attach that to the virtual box. As it stands it's using another interface for qdomyos-zwift via the VM.

**Comment by cagnulein** [comment]
@soccer1886 you don't need the virtualbox anymore! This build takes place of the VM side :) With this you can use dircon with zwift for example

**Comment by soccer1886** [comment]
Is dircon another program or driver I need? When I open the new qdoymos I have no settings to connect to my bluetooth.

**Comment by cagnulein** [comment]
@soccer1886 if you open a prompt on windows, go in the qz folder, qdomyos-zwift.exe -qml 
doesn't work?

**Comment by cagnulein** [comment]
@soccer1886 hah wait, i see something myself too

**Comment by cagnulein** [comment]
@soccer1886 found the issue, i'm building again

**Comment by cagnulein** [comment]
@soccer1886 ok this loads correctly https://github.com/cagnulein/qdomyos-zwift/suites/6274182133/artifacts/224188972

**Comment by soccer1886** [comment]
That did it works like a charm now on PC. Appreciate it.

**Comment by cagnulein** [comment]
@soccer1886 did you try also Dircon and zwift on the same PC?

**Comment by soccer1886** [comment]
Not sure what Dircon is, I have an Echelon connect bike. I originally started qdomyos with the-qml switch and set experimental to wahoo kicker first. Now I just start qdomyos normally, start up Zwift and it picks up my power source, cadence and controllable as Wahoo KICKR. All on the same PC.

I fire up the qdomyos file you just sent me on my pc, set the experimental to wahoo kicker. I then start Zwift and all my connections show up as direct

**Comment by jeremy10tie** [comment]
Hello, I tested the windows version which works well with -qml but the bluetooth works because it searches but finds nothing because it remains blocked on connecting

![Capture](https://user-images.githubusercontent.com/20781131/165707614-fa892917-4c7e-4b15-ba8b-7f20c91d0358.PNG)

**Comment by jeremy10tie** [comment]
Hello, I tested the windows version which works well with -qml and it recognizes well the bike, but on zwift it does not find

**Comment by cagnulein** [comment]
@jeremy10tie did you enable the wahoo dircon as i mentioned above?

**Comment by jeremy10tie** [comment]
> @jeremy10tieavez-vous activé le wahoo dircon comme je l'ai mentionné ci-dessus?

sorry, I forgot to activate it, works perfectly :)

**Comment by cagnulein** [comment]
@jeremy10tie great!

**Comment by cagnulein** [comment]
[debug-Sat_May_7_11_47_09_2022.log](https://github.com/cagnulein/qdomyos-zwift/files/8645556/debug-Sat_May_7_11_47_09_2022.log)
strnage disconnection issue releated to the fact that bluetooth discovery never ends on windows

**Comment by Rossbro2** [comment]
> @soccer1886 ok this loads correctly https://github.com/cagnulein/qdomyos-zwift/suites/6274182133/artifacts/224188972

Looking to set this up on my Windows 11 box today to connect my Echelon EX5 to Zwift 🚴‍♂️ 

However, the above link is not working.  Any ideas?  Thanks in advance!

**Comment by Rossbro2** [comment]
> > @soccer1886 ok this loads correctly https://github.com/cagnulein/qdomyos-zwift/suites/6274182133/artifacts/224188972
> 
> Looking to set this up on my Windows 11 box today to connect my Echelon EX5 to Zwift 🚴‍♂️
> 
> However, the above link is not working. Any ideas? Thanks in advance!

@cagnulein got me going with a link from [facebook](https://www.facebook.com/groups/149984563348738/posts/533586221655235/?comment_id=534746448205879&__cft__[0]=AZUgogVjJxSo-ysQi9Cm398BlzB0rMhR21jaUd0vp9-nR3ee0MHh5l_u3OGEpwVRMwouVBkWoF9M_LaFJKu91OnDcwxLgLxK6TLiTZnlPqe-tgdx6ka2HKGi1os6yQ8SH9WIJMf7uVkBhrUEQ2KDAeIT&__tn__=R]-R).  Thanks!
